README.md
# compojure-api-with-spec

## Pre-requisites
* Java 8
* [Leiningen](https://leiningen.org)

## Usage

```bash
lein ring server
```

## Deploy 

```bash
lein uberjar
java -jar target/server.jar
```

## Test

Run all tests
```bash
lein test
```

`test-refresh` lein plugin is handy if you want to make changes as it re-executes tests affected by the updated code.
```bash

lein test-refresh
```

See how it works from command line with [HTTPie](https:httpie.org).  
```bash
./http_tests
```

## Presentation
[![GitPitch](https://gitpitch.com/assets/badge.svg)](https://gitpitch.com/k2n/compojure-api-with-spec/master?grs=github&t=white)

## Discussion
[![Gitter chat](https://badges.gitter.im/gitterHQ/gitter.png)](https://gitter.im/d3_clojure/Lobby)

Copyright © 2017 Signifier, Inc.

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-09-29
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-09-29
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/compojure-api-with-spec/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/compojure-api-with-spec/compare/0.1.0...0.1.1


test/compojure_api_with_spec/core_test.clj
(ns compojure-api-with-spec.core-test
  (:require [clojure.test :refer :all]
            [compojure-api-with-spec.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


test/compojure_api_with_spec/spec_routes_test.clj
(ns compojure-api-with-spec.spec-routes-test
  (:require [clojure.test :refer :all]
            [clojure.edn :as edn]
            [cheshire.core :as json]
            [compojure-api-with-spec.handler :refer [app]]
            [peridot.core :refer :all]))

(deftest spec-routes-test
  (testing "validation"
    (testing "get"
      (testing "valid email address"
        (is (= 200 (-> (session app)
                       (request "/validation?email-address=foo@example.com")
                       :response
                       :status))))
      (testing "email address without domain"
        (is (= 400 (-> (session app)
                       (request "/validation?email-address=foo")
                       :response
                       :status))))
      (testing "email address without TLD"
        (is (= 400 (-> (session app)
                       (request "/validation?email-address=foo@example")
                       :response
                       :status))))
      (testing "email address with 1 character TLD, which is invalid."
        (is (= 400 (-> (session app)
                       (request "/validation?email-address=foo@example.c")
                       :response
                       :status))))
      (testing "without email-address"
        (is (= 400 (-> (session app)
                       (request "/validation")
                       :response
                       :status))))
      (testing "minus value for age"
        (is (= 400 (-> (session app)
                       (request "/validation?email-address=foo@example.com&current-age=-3")
                       :response
                       :status))))
      (testing "Age 0 is valid"
        (is (= 200 (-> (session app)
                       (request "/validation?email-address=foo@example.com&current-age=0")
                       :response
                       :status))))
      (testing "Age 150 is valid"
        (is (= 400 (-> (session app)
                       (request "/validation?email-address=foo@example.com&current-age=150")
                       :response
                       :status))))
      (testing "Age 48 is categorized as an even number"
        (is (= "あなたの年齢は偶数です"
               (-> (session app)
                   (header "Accept" "application/edn")
                   (request "/validation?email-address=foo@example.com&current-age=48")
                   :response
                   :body
                   slurp
                   edn/read-string
                   :comment))))
      (testing "Age 49 is categorized as an odd number"
        (is (= "あなたの年齢は奇数です"
               (-> (session app)
                   (header "Accept" "application/edn")
                   (request "/validation?email-address=foo@example.com&current-age=49")
                   :response
                   :body
                   slurp
                   edn/read-string
                   :comment)))))
    (testing "post"
      (testing "Accept valid request"
        (is (= "あなたの年齢は奇数です"
               (-> (session app)
                   (header "Accept" "application/edn")
                   (request "/validation"
                            :request-method :post
                            :body-params {:email "foo@example.com"
                                          :age 49})
                   :response
                   :body
                   slurp
                   edn/read-string
                   :comment))))
      (testing "Minus age is rejected"
        (is (= 400
               (-> (session app)
                   (header "Accept" "application/edn")
                   (request "/validation"
                            :request-method :post
                            :body-params {:email "foo@example.com"
                                          :age -49})
                   :response
                   :status))))
      (testing "Incomplete email addresss is rejected"
        (is (= 400
               (-> (session app)
                   (header "Accept" "application/edn")
                   (request "/validation"
                            :request-method :post
                            :body-params {:email "foo@example"
                                          :age 49})
                   :response
                   :status))))))

  (testing "coercion"
    (testing "get"
      (testing "amount-value is passed in as a string query parameter, and 
               conformed as a BigDecimal."
        (is (= 10.23M
               (-> (session app)
                   (header "Accept" "application/edn")
                   (request "/coercion?name=test&amount=10.23")
                   :response
                   :body
                   slurp
                   edn/read-string
                   :amount-value)))))

    (testing "post"
      (testing "amount is passed in as a BigDecimal in EDN, and it is accepted."
        (is (= 12.34M
               (-> (session app)
                   (header "Content-Type" "application/edn")
                   (header "Accept" "application/edn")
                   (request "/coercion"
                            :request-method :post
                            :body-params {:amount 12.34M})
                   :response
                   :body
                   slurp
                   edn/read-string
                   :amount))))

      (testing "amount is passed in as a BigDecimal in JSON, and it is accepted."
        (is (= 12.34
               (-> (session app)
                   (header "Content-Type" "application/json")
                   (header "Accept" "application/json")
                   (request "/coercion"
                            :request-method :post
                            :body-params {:amount 12.34M})
                   :response
                   :body
                   slurp
                   (json/parse-string true)
                   :amount))))

      (testing "Empty body is passed"
        (is (= 400
               (-> (session app)
                   (header "Content-Type" "application/json")
                   (header "Accept" "application/json")
                   (request "/coercion"
                            :request-method :post)
                   :response
                   :status)))))))


project.clj
(defproject compojure-api-with-spec "0.1.0-SNAPSHOT"
  :description "compojure-api v2 and clojure.spec demo"
  :url "https://github.com/k2n/compojure-api-with-spec"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[clj-time                 "0.14.0"]
                 [metosin/compojure-api    "2.0.0-alpha7"]
                 [metosin/spec-tools       "0.3.2"]
                 [org.clojure/clojure      "1.9.0-beta1"]]
  :ring {:handler compojure-api-with-spec.handler/app 
         :async? true
         :nrepl {:start? true}}
  :uberjar-name "server.jar"
  :profiles {:dev {:dependencies [[peridot "0.5.0"]] 
                   :plugins [[lein-ring "0.12.0"]
                             [com.jakemccrary/lein-test-refresh "0.21.1"]]
                   :global-vars {*warn-on-reflection* true}}})


src/compojure_api_with_spec/schema_routes.clj
(ns compojure-api-with-spec.schema-routes
  (:require [compojure.api.sweet :refer [context GET POST]]
            [ring.swagger.json-schema :as rjs]
            [ring.util.http-response :refer [ok]]
            [schema.core :as sc]))

(sc/defschema GreetingResponseSchema 
  {:greeting (rjs/describe sc/Str "Eメールアドレスを含んだ挨拶文")
   :comment (rjs/describe sc/Str "年齢の偶奇数判定")})

(def schema-routes
  (context "/schema" []
    :tags ["schema"]
    :coercion :schema

    (GET "/validate" []
      :summary "Eメールアドレスのフォーマットをチェック"
      :description "Plumatic Schemaを使ったバリデーション。リリース版のため、Descriptionのサポートなど、まだ
                    clojure.specバージョンがサポートできていない機能がフルに利用できる。"
      :query-params [email-address :- (rjs/describe sc/Str "@の前には英数文字、後ろにドメイン、トップレベルドメインが必要。")
                     {current-age :- (rjs/describe sc/Int "任意項目の例。既定値は10") 10}]
      :return GreetingResponseSchema
      (ok {:greeting (str email-address "を登録しました！")
           :comment (if (even? current-age) "あなたの年齢は偶数です" "あなたの年齢は奇数です")}))))


src/compojure_api_with_spec/spec_coercion_routes.clj
(ns compojure-api-with-spec.spec-coercion-routes
  (:require [compojure.api.sweet :refer [context GET POST]]
            [compojure.api.coercion.spec :as spec-coercion]
            [ring.util.http-response :refer [ok]]
            [clojure.spec.alpha :as s]
            [spec-tools.conform :as conform]
            [spec-tools.core :as st]
            [spec-tools.spec :as spec]))

(s/def ::name spec/string?)
(s/def ::amount (assoc spec/bigdec?
                       :json-schema/type "string"
                       :description "整数または小数を文字列として入力"))
(s/def ::amount-type spec/string?)

(defn string->bigdec [_ x]
  (cond
    (bigdec? x) x
    (string? x) (try
                  (java.math.BigDecimal. x)
                  (catch NumberFormatException e
                    ::s/invalid))
    :else ::s/invalid))

(def string-conforming
  (st/type-conforming
   (merge
    (assoc conform/string-type-conforming :bigdec string->bigdec)
    conform/strip-extra-keys-type-conforming)))

(def json-conforming
  (st/type-conforming
   (merge
    (assoc conform/json-type-conforming :bigdec string->bigdec)
    conform/strip-extra-keys-type-conforming)))

(def custom-spec-coercion
  (spec-coercion/create-coercion
   (-> (assoc-in spec-coercion/default-options
                 [:body :formats "application/json"] json-conforming)
       (assoc-in [:string :default] string-conforming))))

(def spec-coercion-routes
  (context "/coercion" []
    :tags ["coercion"]
    :coercion custom-spec-coercion

    (GET "/" []
      :query-params [name :- ::name amount :- ::amount]
      (ok {:greet (str "Hello, " name "!")
           :amount-type (str (type amount))
           :amount-value amount}))

    (POST "/" []
      :body [{:keys [amount]} (s/keys :req-un [::amount])]
      :return (s/keys :req-un [::amount-type ::amount])
      (ok {:amount-type (str (type  amount))
           :amount amount}))))


src/compojure_api_with_spec/spec_validation_routes.clj
(ns compojure-api-with-spec.spec-validation-routes
  (:require [compojure.api.sweet :refer [context GET POST]]
            [ring.util.http-response :refer [ok]]
            [clojure.spec.alpha :as s]
            [spec-tools.spec :as spec]))

(def email-regex #"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,63}$")
(def max-age 140)

(s/def ::greeting (assoc spec/string?
                         :description "このように注釈を定義することができる"))
(s/def ::comment  (assoc spec/string?
                         :description "年齢の偶奇数を判定します"))
(s/def ::response (s/keys :req-un [::greeting ::comment]))
(s/def ::email (s/and spec/string? #(re-matches email-regex %)))
(s/def ::age  (s/and spec/int?
                     #(> max-age %) #((complement neg?) %)))

(defn validation-handler
  [email-address current-age]
  {:pre [(s/valid? ::email email-address)
         (s/valid? ::age current-age)]
   :post [(s/valid? ::response %)]}
  {:greeting (str email-address "を登録しました！")
   :comment (if (even? current-age)
              "あなたの年齢は偶数です"
              "あなたの年齢は奇数です")})

(def spec-validation-routes
  (context "/validation" []
    :coercion :spec
    :tags ["validation"]

    (GET "/" []
      :summary "クエリーパラメタの入力チェック"
      :description "正規表現によるEメールアドレスのフォーマットチェックと、
                   整数の範囲によるバリデーションのデモ"
      :query-params [email-address :- ::email
                     {current-age :- ::age  18}]
      :return ::response
      (ok (validation-handler email-address current-age)))

    (POST "/" []
      :summary "Bodyの入力チェック"
      :return ::response
      :body [{:keys [email age]} (s/keys :req-un [::email ::age])]
      (ok (validation-handler email age)))))


src/compojure_api_with_spec/spec_routes.clj
(ns compojure-api-with-spec.spec-routes
  (:require [compojure.api.sweet :refer [context GET POST]]
            [compojure.api.coercion.spec :as spec-coercion]
            [ring.util.http-response :refer [ok]]
            [clojure.spec.alpha :as s]
            [spec-tools.conform :as conform]
            [spec-tools.core :as st]
            [spec-tools.spec :as spec]))

(def email-regex #"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,63}$")

(s/def ::greeting spec/string?)
(s/def ::comment  spec/string?)
(s/def ::email (s/and spec/string? #(re-matches email-regex %)))
(s/def ::age (s/and spec/int? #(> 140 %) #((complement neg?) %)))
(s/def ::name spec/string?)
(s/def ::amount spec/bigdec?)

(defn string->bigdec [_ x]
  (cond
    (bigdec? x) x
    (string? x) (try
                  (java.math.BigDecimal. x)
                  (catch NumberFormatException e
                    ::s/invalid))
    :else ::s/invalid))

(def string-conforming
  (st/type-conforming
   (merge
    (assoc conform/string-type-conforming :bigdec string->bigdec)
    conform/strip-extra-keys-type-conforming)))

(def json-conforming
  (st/type-conforming
   (merge
    (assoc conform/json-type-conforming :bigdec string->bigdec)
    conform/strip-extra-keys-type-conforming)))

(def custom-spec-coercion
  (spec-coercion/create-coercion
   (-> (assoc-in spec-coercion/default-options [:body :formats "application/json"] json-conforming)
       (assoc-in [:string :default] string-conforming))))

(def spec-validation-routes
  (context "/validation" []
    :coercion :spec
    :tags ["validation"]

    (GET "/" []
      :summary "クエリーパラメタの入力チェック"
      :description "正規表現によるEメールアドレスのフォーマットチェックと、整数の範囲によるバリデーションのデモ"
      :query-params [email-address :- ::email
                     {current-age :- ::age  18}]
      :return (s/keys :req-un [::greeting ::comment])
      (ok {:greeting (str email-address "を登録しました！")
           :comment (if (even? current-age) "あなたの年齢は偶数です" "あなたの年齢は奇数です")}))

    (POST "/" []
      :summary "Bodyの入力チェック"
      :body [{:keys [email age]} (s/keys :req-un [::email ::age])]
      (ok {:greeting (str email "を登録しました！")
           :comment (if (even? age) "あなたの年齢は偶数です" "あなたの年齢は奇数です")}))))

(def spec-coercion-routes
  (context "/coercion" []
    :tags ["coercion"]
    :coercion custom-spec-coercion

    (GET "/" []
      :query-params [name :- ::name amount :- ::amount]
      (ok {:greet (str "Hello, " name "!")
           :amount-type (str (type amount))
           :amount-value amount}))

    (POST "/" []
      :body [amount (s/keys :req-un [::amount])]
      (ok {:greet "Hello!"
           :amount-type (str (type (:amount amount)))
           :amount-value amount}))))
